Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,391 @@
+#include "Projectiles.h"
+#include "../Helpers.h"
+
+
+Projectile::Projectile(size_t id) : IMovable(id)
+{
+    SetPassable(D_ALL, Passable::AIR);
+    tickSpeed = 0.25;
+    damage = 0;
+    SetSprite("icons/projectiles.dmi");
+    v_level = 5;
+    got_bored_ = 0;
+    set_target_ = 0;
+}
+void Projectile::Process()
+{
+   if(set_target_ == 0)
+   {
+      return;
+   }
+   Dir step = CalculateTrajectoryMove();
+   Rotate(step);
+   if(!CheckPassable())
+   {
+       this->Delete();
+       return;
+   }
+   MainMove(); 
+}
+
+bool Projectile::CheckPassable()
+{
+    id_ptr_on<Projectile> p = this->GetId();
+    PassableLevel loc = GetPassable(GetDir());
+    if (loc != Passable::FULL)
+    {
+        SetPassable(GetDir(), Passable::FULL);
+    }
+    if (!CanPass(owner->GetPassable(GetDir()), passable_level))
+    {
+        owner->Bump(p);
+        if (loc != Passable::FULL)
+        {
+            SetPassable(GetDir(), loc);
+        }
+        return false;
+    }
+    if (loc != Passable::FULL)
+    {
+        SetPassable(GetDir(), loc);
+    }
+
+    auto neighbour = owner->GetNeighbour(GetDir());
+    if (   !CanPass(neighbour->GetPassable(D_ALL), passable_level)
+        || !CanPass(neighbour->GetPassable(helpers::revert_dir(GetDir())), passable_level))
+    {
+        neighbour->Bump(p);
+        return false;
+    }
+    
+    return true;
+}
+void Projectile::Bump(id_ptr_on<IMovable> item)
+{
+    id_ptr_on<Projectile> p = this->GetId();
+    item->Bump(p);
+    this->Delete();
+   
+//TODO maybe it can apply froce depending on bullet speed but that should be considered        ApplyForce(DirToVDir[m->dMove]);
+}
+void Projectile::MakeMovementPattern(VDir target,id_ptr_on<Human> someone)
+{
+    set_target_=1;
+    target_ = target;
+    int x = target.x;
+    int y = target.y;
```

Comment:
Your previous approach with an array of next moves was much better than current one. As far as I can see from the unsync dumps it appears because a laser projectile is on different tiles in different game instances. So seems like the issue in the movement calculation.
Current code is totally unreadable.
If your issue was with `std::vector` and `KV_SAVEBLE` then it is just needed to make serialization functions for that variable type. (It is needed to implement `operator>>` and `operator<<` for `std::vector<Dir>`, there should be some examples in the code for other types).